mcs/class/System/System.Net/HttpWebRequest.cs

```diff
@@ -853,71 +882,61 @@ public override IAsyncResult BeginGetRequestStream (AsyncCallback callback, obje
 			if (!sendChunked && transferEncoding != null && transferEncoding.Trim () != "")
 				throw new ProtocolViolationException ("SendChunked should be true.");
 
-			lock (locker)
-			{
+			WebOperation operation;
+			lock (locker) {
 				if (getResponseCalled)
 					throw new InvalidOperationException ("The operation cannot be performed once the request has been submitted.");
 
-				if (asyncWrite != null) {
+				operation = currentOperation;
+				if (operation != null && operation.WriteStream != null)
+					return operation.WriteStream;
+
+				if (Interlocked.CompareExchange (ref nestedRequestStream, 1, 0) != 0)
 					throw new InvalidOperationException ("Cannot re-call start of asynchronous " +
 								"method while a previous call is still in progress.");
-				}
-	
-				asyncWrite = new WebAsyncResult (this, callback, state);
+
 				initialMethod = method;
-				if (haveRequest) {
-					if (writeStream != null) {
-						asyncWrite.SetCompleted (true, writeStream);
-						asyncWrite.DoCallback ();
-						return asyncWrite;
-					}
-				}
-				
-				gotRequestStream = true;
-				WebAsyncResult result = asyncWrite;
-				if (!requestSent) {
-					requestSent = true;
-					redirects = 0;
-					servicePoint = GetServicePoint ();
-					abortHandler = servicePoint.SendRequest (this, connectionGroup);
+
+				if (operation == null) {
+					gotRequestStream = true;
+					operation = SendRequest (false, null, cancellationToken);
 				}
-				return result;
+			}
+
+			try {
+				return await operation.GetRequestStream ().ConfigureAwait (false);
+			} finally {
+				nestedRequestStream = 0;
```

Comment:
Could you add a comment why
Can this be racy?

Reply:
I believe this variable does not need to be reset, but need to check this.
**MARTIN TODO**